### PR TITLE
Add system_timer_wait_cycles

### PR DESCRIPTION
**Pull request by jamesadevine**

### CODE REVIEW

**Comment by mmoskal** [comment]
how about adding system_timer_calibrate() and using the wait_cycles thing in default wait_us?

**Comment by mmoskal** [comment]
this way, if target init calls calibrate(), then it get precise timing

**Comment by mmoskal** [comment]
maybe the 16 bit fixed point is overkill and an integer is fine, especially if the delay loop was all in assembly, like you had it initially (sorry for confusion)

**Comment by jamesadevine** [comment]
@mmoskal Completed revisions as per our discussion:

* swap wait_ms / wait_us to use `uint32_t` rather than `uint64_t`.
* drop fixed point computation.
* include automatic calculate of cycles per microsecond.

File: inc/driver-models/Timer.h [hunk 1]
```diff
@@ -265,29 +265,46 @@ namespace codal
...
+      * An auto calibration method that uses the hardware timer to compute the number of cycles
+      * per us.
+      *
+      * The result of this method is then used to compute accurate us waits using instruction counting in system_timer_wait_us.
+      *
+      * If this method is not called, a less accurate timer implementation is used in system_timer_wait_us.
+      *
+      * @return DEVICE_OK on success, and DEVICE_NOT_SUPPORTED if no system_timer yet exists.
+      */
+    int system_timer_calibrate_cycles();
```
Comment (mmoskal) [inline inc/driver-models/Timer.h]: call it from Timer ctor?

Comment (jamesadevine) [reply inc/driver-models/Timer.h]: Good idea, done.

**Pull request merged**